### PR TITLE
Undo for git-rename-branch

### PR DESCRIPTION
**Pull request by kevgo**
@ricmatsui @charlierudolph @allewun 

* adds the undo path for `git-rename-branch`
* splits up the `push_branch` helper into one that creates a new remote branch (`create_tracking_branch`), and one that merely pushes an existing branch (`push_branch`)

implements #519 and #472 
unblocks #664

### CODE REVIEW

**Comment by kevgo** [comment]
bump

File: src/helpers/configuration_helpers.sh [hunk 1]
```diff
@@ -198,3 +198,17 @@ function store_configuration {
...
+  echo "remove_perennial_branch $branch_name"
+}
+
+
+function undo_steps_for_remove_perennial_branch {
+  local branch_name=$1
+  echo "add_perennial_branch $branch_name"
+}
+
+
```
Comment (charlierudolph) [inline src/helpers/configuration_helpers.sh]: extra whitespace can be removed

File: src/helpers/git_helpers/feature_branch_helpers.sh [hunk 2]
```diff
@@ -65,3 +65,14 @@ function is_perennial_branch {
     echo false
   fi
 }
+
+
+function undo_steps_for_create_and_checkout_feature_branch {
+  local current_branch=$(get_current_branch_name)
+  local branch_to_create="$1"
```
Comment (charlierudolph) [inline src/helpers/git_helpers/feature_branch_helpers.sh]: We have always put arguments first. Also I believe the quotes are unnecessary

File: src/helpers/configuration_helpers.sh (hunk 1 above)
Comment (kevgo) [inline src/helpers/configuration_helpers.sh]: done

File: src/helpers/git_helpers/feature_branch_helpers.sh (hunk 2 above)
Comment (kevgo) [inline src/helpers/git_helpers/feature_branch_helpers.sh]: done

**Comment by kevgo** [comment]
All comments addressed

File: documentation/commands/git-new-pull-request.md [hunk 3]
```diff
@@ -12,10 +12,12 @@ git new-pull-request
...
 
-Opens a browser window to the form for a new pull request page
-for the current branch on the code hosting platform.
+Syncs the current branch
+and opens a browser window to the new pull request page of your repository.
 
-The pull request only shows the changes made against the immediate parent branch.
+The form is pre-populated for the current branch
+so that the pull request only shows the changes made
+against the immediate parent branch.
```
Comment (charlierudolph) [inline documentation/commands/git-new-pull-request.md]: These changes are unrelated to this PR. Looks like they were meant for another branch

File: src/helpers/git_helpers/branch_helpers.sh [hunk 4]
```diff
@@ -187,17 +190,24 @@ function remote_only_merged_branches {
...
 
 
-function undo_steps_for_create_and_checkout_feature_branch {
-  local current_branch=$(get_current_branch_name)
+function undo_steps_for_create_and_checkout_branch {
   local branch_to_create="$1"
 
+  local current_branch=$(get_current_branch_name)
   echo "checkout $current_branch"
+
```
Comment (charlierudolph) [inline src/helpers/git_helpers/branch_helpers.sh]: I think this should follow a similar style as `undo_steps_for_create_and_checkout_feature_branch`.
All the variable definitions, then a empty line, then all the echos.

File: src/git-rename-branch [hunk 5]
```diff
@@ -85,6 +83,8 @@ function steps {
   echo delete_parent_entry "$target_branch_name"
   echo store_parent_branch "$destination_branch_name" "$parent_branch_name"
   echo "delete_all_ancestor_entries"
+
+  echo "delete_local_branch '$target_branch_name' force"
```
Comment (charlierudolph) [inline src/git-rename-branch]: Just curios, why did this have to be moved?

File: documentation/commands/git-new-pull-request.md (hunk 3 above)
Comment (kevgo) [inline documentation/commands/git-new-pull-request.md]: Good catch. I have extracted this to a different branch.

File: src/git-rename-branch (hunk 5 above)
Comment (kevgo) [inline src/git-rename-branch]: Undoing this runs operations that require this branch to exist locally. Hence, it has to be created first thing in the undo operation. Hence it has to be deleted last.

File: src/helpers/git_helpers/branch_helpers.sh (hunk 4 above)
Comment (kevgo) [inline src/helpers/git_helpers/branch_helpers.sh]: Done

**Comment by kevgo** [comment]
All comments addressed

**Comment by charlierudolph** [comment]
:ship:
